File: packages/native/cli/N8Bridge.cpp
#include "N8Bridge.hpp"

#include <atomic>
#include <chrono>
#include <csignal>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <deque>
#include <memory>
#include <string>
#include <thread>
#include <vector>

#include "host/input/MidiIo.hpp"
#include "host/n8/Edio.hpp"
#include "host/n8/WjwwoodSerialPort.hpp"

namespace retroplug {

namespace {

// Ctrl-C stop flag. SIGINT is async-signal-limited, so the handler only flips this atomic; the loop
// polls it. One process, one bridge, so a TU-local is the right mechanism.
std::atomic<bool> g_stop{false};
void onSigint(int) { g_stop.store(true, std::memory_order_relaxed); }

void printUsage() {
    std::puts(
        "usage: retroplug-cli n8-bridge [options]\n"
        "\n"
        "  Pipe live MIDI input to a physical Everdrive N8 Pro over USB (running EverMIDI),\n"
        "  so a controller / DAW plays the real NES. One-way (host -> cart). Ctrl-C to stop.\n"
        "\n"
        "options:\n"
        "  --list               list MIDI inputs + serial ports (the N8 is tagged) and exit\n"
        "  --midi-in <name>     use this MIDI input (default: all hardware inputs)\n"
        "  --serial <port>      use this serial port (default: auto-detect the N8, VID:PID 38df:0017)\n"
        "  --lookahead-ms <N>   delay each message by N ms and release on a timed schedule (default: 0,\n"
        "                       immediate forward - the MIDI source's own timing is authoritative)\n"
        "  -h, --help           show this help");
}

int runList() {
    std::puts("MIDI inputs:");
    {
        MidiIo midi;
        const std::vector<std::string> inputs = midi.listInputs();
        if (inputs.empty()) {
            std::puts("  (none)");
        } else {
            for (const std::string& name : inputs) std::printf("  %s\n", name.c_str());
        }
    }
    std::puts("\nSerial ports:");
    const std::vector<N8PortInfo> ports = listSerialPorts();
    if (ports.empty()) {
        std::puts("  (none)");
    } else {
        for (const N8PortInfo& p : ports) {
            std::string line = "  " + p.port;
            if (!p.description.empty()) line += "  (" + p.description + ")";
            if (p.isN8) line += "  [Everdrive N8 Pro]";
            std::puts(line.c_str());
        }
    }
    return 0;
}

}  // namespace

int runN8Bridge(int argc, char** argv) {
    // Parse argv[2..] (argv[1] == "n8-bridge").
    bool        list = false;
    std::string midiIn;
    std::string serialPortName;
    int         lookaheadMs = 0;

    for (int i = 2; i < argc; ++i) {
        const char* a = argv[i];
        auto next = [&](const char* flag) -> const char* {
            if (i + 1 >= argc) {
                std::fprintf(stderr, "error: %s needs a value\n", flag);
                std::exit(2);
            }
            return argv[++i];
        };
        if (std::strcmp(a, "--list") == 0) {
            list = true;
        } else if (std::strcmp(a, "--midi-in") == 0) {
            midiIn = next("--midi-in");
        } else if (std::strcmp(a, "--serial") == 0) {
            serialPortName = next("--serial");
        } else if (std::strcmp(a, "--lookahead-ms") == 0) {
            lookaheadMs = std::atoi(next("--lookahead-ms"));
            if (lookaheadMs < 0) lookaheadMs = 0;
        } else if (std::strcmp(a, "-h") == 0 || std::strcmp(a, "--help") == 0) {
            printUsage();
            return 0;
        } else {
            std::fprintf(stderr, "error: unknown option '%s' (try --help)\n", a);
            return 2;
        }
    }

    if (list) return runList();

    // Resolve the serial port (explicit --serial, else auto-detect the N8).
    std::string portName = serialPortName.empty() ? findN8Port() : serialPortName;
    if (portName.empty()) {
        std::fprintf(stderr,
                     "error: no Everdrive N8 Pro found (VID:PID 38df:0017). Plug it in, or pass\n"
                     "       --serial <port>. Run 'retroplug-cli n8-bridge --list' to see ports.\n");
        return 1;
    }

    std::unique_ptr<WjwwoodSerialPort> port;
    try {
        port = std::make_unique<WjwwoodSerialPort>(portName);
    } catch (const std::exception& e) {
        std::fprintf(stderr, "error: cannot open serial port %s: %s\n", portName.c_str(), e.what());
        return 1;
    }

    Edio edio(*port);
    try {
        edio.connect();  // handshake; throws on a bad / absent reply
    } catch (const std::exception& e) {
        std::fprintf(stderr, "error: N8 handshake failed on %s: %s\n", portName.c_str(), e.what());
        return 1;
    }
    std::printf("Connected to Everdrive N8 Pro on %s\n", portName.c_str());

    MidiIo midi;
    if (!midi.open("RetroPlug N8")) {
        std::fprintf(stderr, "error: no MIDI system available\n");
        return 1;
    }
    midi.setInputSelection(midiIn);  // "" = all hardware inputs
    if (midiIn.empty())
        std::puts("MIDI input: all hardware inputs");
    else
        std::printf("MIDI input: %s\n", midiIn.c_str());
    if (lookaheadMs > 0) std::printf("Lookahead: %d ms (timed release)\n", lookaheadMs);
    std::puts("Bridging - Ctrl-C to stop.");

    std::signal(SIGINT, onSigint);

    // Forward one message to the cart FIFO. Returns false if the serial link died (stop the loop).
    std::uint64_t msgCount = 0, byteCount = 0;
    auto forward = [&](const std::vector<std::uint8_t>& bytes) -> bool {
        if (bytes.empty()) return true;
        try {
            edio.fifoWR(bytes);
        } catch (const std::exception& e) {
            std::fprintf(stderr, "\nerror: serial write failed (N8 unplugged?): %s\n", e.what());
            return false;
        }
        ++msgCount;
        byteCount += bytes.size();
        return true;
    };

    using clock = std::chrono::steady_clock;
    struct Pending { clock::time_point due; std::vector<std::uint8_t> bytes; };
    std::deque<Pending> queue;  // lookahead>0: constant delay -> arrival order == release order (FIFO)

    std::vector<MidiIo::Message> scratch;
    clock::time_point lastStatus = clock::now();
    bool alive = true;

    while (!g_stop.load(std::memory_order_relaxed) && alive) {
        const clock::time_point now = clock::now();

        midi.poll(scratch);
        for (MidiIo::Message& m : scratch) {
            if (m.bytes.empty()) continue;
            if (lookaheadMs > 0) {
                queue.push_back({now + std::chrono::milliseconds(lookaheadMs), std::move(m.bytes)});
            } else if (!forward(m.bytes)) {
                alive = false;
                break;
            }
        }

        // Release any due messages (lookahead path).
        while (alive && !queue.empty() && queue.front().due <= now) {
            if (!forward(queue.front().bytes)) { alive = false; break; }
            queue.pop_front();
        }

        // Throttled status line (in place).
        if (now - lastStatus >= std::chrono::milliseconds(500)) {
            std::printf("\r  forwarded %llu messages / %llu bytes   ",
                        static_cast<unsigned long long>(msgCount),
                        static_cast<unsigned long long>(byteCount));
            std::fflush(stdout);
            lastStatus = now;
        }

        std::this_thread::sleep_for(std::chrono::microseconds(250));
    }

    std::printf("\nStopped. Forwarded %llu messages / %llu bytes.\n",
                static_cast<unsigned long long>(msgCount), static_cast<unsigned long long>(byteCount));
    midi.close();
    return 0;
}

}  // namespace retroplug

File: packages/retroplug/cli/sessions/n8-bridge.ts
// `retroplug-cli n8-bridge` / `n8-sync` - stream live MIDI to a physical Everdrive N8 Pro over USB, entirely
// in TS (these replaced the native subcommands). `n8-bridge` forwards raw MIDI to the cart FIFO (for the
// EverMIDI ROM); `n8-sync` translates a MIDI clock/transport into risa host-sync bytes (via the pure-TS
// RisaSyncTranslator). Both are long-running: they open the serial (createN8) + MIDI (createMidiClient)
// facets, call keepAlive(), and run an event-driven poll loop (setInterval) until Ctrl-C - the native
// launcher pumps it (see cli/main.cpp).
import type { CliTool } from "../tools";
import type { Session } from "../session";
import { keepAlive, exitProcess } from "../session";
import { createSerialClient, createMidiClient } from "../../src/realBackend";
import { createN8, RisaSyncTranslator, type SerialPortInfo } from "../../src/n8";

// txiki provides this libuv-backed global at runtime (deps/.../txiki timers); the TS lib config omits it.
declare function setInterval(handler: () => void, ms: number): number;

const flag = (args: string[], name: string): string | undefined => {
  const i = args.indexOf(name);
  return i >= 0 ? args[i + 1] : undefined;
};

// The port to talk to: an explicit --serial, else the first attached N8 (USB VID:PID 38df:0017).
function pickPort(ports: SerialPortInfo[], explicit: string | undefined): string {
  if (explicit) return explicit;
  const n8 = ports.find((p) => p.isN8);
  if (!n8)
    throw new Error(
      "no Everdrive N8 Pro found (VID:PID 38df:0017). Plug it in, or pass --serial <port>.\n" +
        (ports.length ? `serial ports: ${ports.map((p) => p.port).join(", ")}` : "no serial ports detected"),
    );
  return n8.port;
}

function printInputs(): void {
  const inputs = createMidiClient().listInputs();
  if (inputs.length === 0) {
    console.log("no MIDI inputs found");
    return;
  }
  console.log("MIDI inputs:");
  for (const n of inputs) console.log(`  ${n}`);
}

// The shared bridge loop: poll MIDI, transform each message to output bytes, forward to the cart FIFO with an
// optional constant-delay lookahead (arrival order == release order). `transform` is the ONLY difference
// between the raw bridge and the risa-sync bridge. Long-running: sets keepAlive() and returns; the native
// pump drives the setInterval poll until Ctrl-C.
function runBridge(args: string[], clientName: string, transform: (bytes: Uint8Array, out: number[]) => void): void {
  const lookaheadMs = Math.max(0, Number(flag(args, "--lookahead-ms") ?? 0) | 0);

  const serial = createSerialClient();
  const port = serial.open(pickPort(serial.listPorts(), flag(args, "--serial")));
  const n8 = createN8(port);
  n8.connect(); // throws if the N8 doesn't answer the handshake

  const midi = createMidiClient();
  const midiIn = flag(args, "--midi-in") ?? "";
  if (!midi.open(clientName, midiIn)) throw new Error("no MIDI system available");

  keepAlive(); // opt into the native run-until-Ctrl-C pump

  const queue: { due: number; bytes: number[] }[] = [];
  let msgCount = 0;
  let byteCount = 0;
  const forward = (bytes: number[]): void => {
    n8.edio.fifoWR(Uint8Array.from(bytes));
    msgCount++;
    byteCount += bytes.length;
  };

  setInterval(() => {
    const now = Date.now();
    for (const bytes of midi.poll()) {
      const out: number[] = [];
      transform(bytes, out);
      if (out.length === 0) continue; // non-transport MIDI (sync) yields nothing
      if (lookaheadMs > 0) queue.push({ due: now + lookaheadMs, bytes: out });
      else forward(out);
    }
    while (queue.length > 0 && queue[0].due <= now) forward(queue.shift()!.bytes);
  }, 1);

  setInterval(() => console.log(`forwarded ${msgCount} messages / ${byteCount} bytes`), 2000);

  console.log(
    `bridging ${midiIn || "all MIDI inputs"} -> Everdrive N8 on ${port.port}` +
      `${lookaheadMs ? ` (+${lookaheadMs}ms lookahead)` : ""}. Ctrl-C to stop.`,
  );
}

const COMMON_FLAGS = [
  "  --list             list the available MIDI input ports and exit",
  "  --midi-in <name>   forward from this MIDI input (default: all hardware inputs)",
  "  --serial <port>    use this serial port (default: auto-detect the N8, VID:PID 38df:0017)",
  "  --lookahead-ms <N> delay every message by N ms (constant-delay smoothing; default 0 = immediate)",
];

const BRIDGE_HELP = [
  "usage: retroplug-cli n8-bridge [--list] [--midi-in <name>] [--serial <port>] [--lookahead-ms <N>]",
  "",
  "  Stream live MIDI straight to a physical Everdrive N8 Pro's cart FIFO over USB, so a controller / DAW",
  "  plays the real NES (for the EverMIDI ROM). Runs until Ctrl-C.",
  "",
  ...COMMON_FLAGS,
].join("\n");

const SYNC_HELP = [
  "usage: retroplug-cli n8-sync [--list] [--midi-in <name>] [--serial <port>] [--lookahead-ms <N>]",
  "",
  "  Turn an incoming MIDI clock/transport (Start/Continue/Stop + 24-PPQN clock, optional Song Position)",
  "  into risa's host-sync protocol on a physical N8, so a risa cart plays locked to the DAW. Runs until",
  "  Ctrl-C. Load a risa ROM first (retroplug-cli n8-load).",
  "",
  ...COMMON_FLAGS,
].join("\n");

export const n8BridgeTool: CliTool = {
  name: "n8-bridge",
  summary: "stream live MIDI to a physical Everdrive N8 Pro over USB",
  help: BRIDGE_HELP,
  longRunning: true,
  run(_s: Session, args: string[]): void {
    if (args.includes("--list")) {
      printInputs();
      exitProcess(0);
      return;
    }
    runBridge(args, "RetroPlug N8", (bytes, out) => {
      for (const b of bytes) out.push(b); // raw forward
    });
  },
};

export const n8SyncTool: CliTool = {
  name: "n8-sync",
  summary: "drive risa host-sync on a physical Everdrive N8 Pro from a MIDI clock",
  help: SYNC_HELP,
  longRunning: true,
  run(_s: Session, args: string[]): void {
    if (args.includes("--list")) {
      printInputs();
      exitProcess(0);
      return;
    }
    const translator = new RisaSyncTranslator();
    runBridge(args, "RetroPlug N8 Sync", (bytes, out) => translator.onMessage(bytes, out));
  },
};
